pkg/operator/events/recorder.go

```diff
@@ -0,0 +1,92 @@
+package events
+
+import (
+	"fmt"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+type RecorderSourceFunc func() (RecorderSource, error)
+
+type Recorder interface {
+	CreateEvent(reason, message string) error
+	CreateEventWarning(reason, message string) error
+}
+
+type RecorderSource interface {
+	GroupVersionKind() schema.GroupVersionKind
+	metav1.ObjectMetaAccessor
+}
+
+func NewRecorder(client corev1client.EventInterface, sourceComponent string, sourceFunc RecorderSourceFunc) Recorder {
```

Comment:
why the RecorderSourceFunc and not RecorderSource directly? The func without arguments is like a constant for the record, isn't it?

Reply:
yeah, good point

Reply:
@sttts actually, `RecorderSourceFunc` returns the RecorderSource and error and it allows things like `DeploymentFromEnvRecorderSource` to be used as argument. I don't see how to refactor this to loose the RecorderSourceFunc ?